server/llm/openai_assistant.py

```diff
@@ -45,41 +50,56 @@ def register_new_context(self, new_text: str, metadata: list[str] = None):
         """Registers new context (usually a transcription line)."""
         content = self._compile_ctx_content(new_text, metadata)
         user_msg = ChatCompletionUserMessageParam(content=content, role="user")
-        self._context.append(user_msg)
+        with self._lock:
+            self._context.append(user_msg)
 
-    def query(self, custom_query: str = None) -> str:
+    async def query(self, custom_query: str = None) -> str:
         """Submits a query to OpenAI with the stored context if one is provided.
         If a query is not provided, uses the default."""
-        if len(self._context) == 0:
-            raise NoContextError()
+        with self._lock:
+            if len(self._context) == 0:
+                raise NoContextError()
 
         query = self._default_prompt
 
         if custom_query:
             query = ChatCompletionSystemMessageParam(
                 content=custom_query, role="system")
 
-        messages = self._context + [query]
+        with self._lock:
+            messages = self._context + [query]
 
         try:
-            res = self._client.chat.completions.create(
-                model=self._model_name,
-                messages=messages,
-                temperature=0,
-            )
-            for choice in res.choices:
-                reason = choice.finish_reason
-                if reason == "stop" or reason == "length":
-                    answer = choice.message.content
-                    return answer
-            raise Exception("No usable choice found in OpenAI response: %s", res.choices)
+            loop = asyncio.get_event_loop()
+            future = loop.run_in_executor(
```

Comment:
This change runs this otherwise blocking IO operation in the default `ThreadPoolExecutor()` to avoid blocking the main thread.